### PR TITLE
Fixes FlowContainer inappropriately adding it's padding

### PR DESCRIPTION
**Pull request by RemieRichards**
Fixes ```FlowContainer``` inappropriately adding it's padding, padding is between elements, not between an element and the ```FlowContainer```'s edge.

* ```FlowContainer``` ```Padding``` is now called ```Spacing``` by Request of Peppy, in regards to potentially moving ```PaddingContainer```'s ```Padding``` functionality down to ```Container``` (Along with new properties like Margin)
* Puts ```TestCaseScrollableFlow```  in a ``` PaddingContainer``` so that it visually looks the same, without relying on the above pseudo-behaviour

To test this I took ```TestCaseScrollableFlow``` and <del>butchered it</del> changed it's ```Padding```(now ```Spacing```) to be ```Vector2(10, 10)``` instead of 1,1, so I could actually *see* changes.
Before fixing:
OldBehaviour

After fixing:
NewBehaviour

As you can see my by cursor wiggle, the inappropriate padding is gone, this is almost guaranteed to cause some minor visual offsets elsewhere in the game, but more fool those parts for relying on this pseudo-behaviour. One such visual offset is visible in the "After Fixing" gif, the nice scrollable flow of boxes is now squished against the test bar, so I've offset it's position by 5 pixels using the now Container-level ```Padding```, so it looks identical to how it did before, just correct (it's also back to ```Vector(1, 1)``` again)

This is sort of a companion to https://github.com/ppy/osu-framework/pull/132, as both PRs are a potential fix for issues i had during implementation, however this is a fix so it's almost certainly required, where as the other one is more of a "it'd be nice to have this" (From talks with Peppy that PR should work once I allow ```AutoSizeContainer``` to take padding into account of it's bounds) 

Todo:
- [x] Update after https://github.com/ppy/osu-framework/pull/132

### CODE REVIEW

File: osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs [hunk 1]
```diff
@@ -36,7 +36,8 @@ public override void Reset()
                         {
                             LayoutDuration = 100,
                             LayoutEasing = EasingTypes.Out,
-                            Padding = new Vector2(1, 1),
+                            Spacing = new Vector2(1, 1),
+							Position = new Vector2(5, 5),
```
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: tabs are tabs not spaces

Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: also just setting the position to 5,5 is gonna make the content go off-screen, right?

File: osu.Framework/Graphics/Containers/FlowContainer.cs [hunk 2]
```diff
@@ -52,16 +52,16 @@ public Vector2 MaximumSize
             }
         }
 
-        Vector2 padding;
+		Vector2 spacing;
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: wrong tabs

File: osu.Framework/Graphics/Containers/FlowContainer.cs [hunk 3]
```diff
@@ -52,16 +52,16 @@ public Vector2 MaximumSize
             }
         }
 
-        Vector2 padding;
+		Vector2 spacing;
 
-        public Vector2 Padding
+        public Vector2 Spacing
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: Add xmldoc explaining that this is the spacing placed between elements only.

File: osu.Framework/Graphics/Containers/FlowContainer.cs [hunk 4]
```diff
@@ -123,15 +123,15 @@ protected override void UpdateLayout()
 
                             if (Direction != FlowDirection.HorizontalOnly && current.X + size.X > max.X)
                             {
-                                current.X = Math.Max(0, Padding.X);
-                                current.Y += rowMaxHeight;
+								current.X = 0; //We use this for the above conditional, then discard
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: this comment doesn't make sense (and  is probably not even necessary)

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 5]
```diff
@@ -180,7 +180,7 @@ private float getPositionAt(int index)
                 if (index < text.Length)
                     return textFlow.Children.ElementAt(index).Position.X + textFlow.Position.X;
                 var d = textFlow.Children.ElementAt(index - 1);
-                return d.Position.X + d.Size.X + textFlow.Padding.X + textFlow.Position.X;
+                return d.Position.X + d.Size.X + textFlow.Spacing.X + textFlow.Position.X;
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: did you check for regressions from these changes?

File: osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs (hunk 1 above)
Comment (RemieRichards) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: 5,5 just puts it back to where it was (visually) before the fix, so it changes nothing content wise.
(if you look at the "fixed" gif above, it hugs the test buttons, where as before it doesn't, the 5,5 is just a visual niceness for the test)

File: osu.Framework/Graphics/Containers/FlowContainer.cs (hunk 2 above)
Comment (RemieRichards) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: Yeah Xamarin likes to forget my preferences, I'll sort it.

File: osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs (hunk 1 above)
Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: Yes, but you're adding it to a relativeX container, which means 5 pixels of the container is now out-of-bounds to the right?

File: osu.Framework/Graphics/Containers/FlowContainer.cs (hunk 4 above)
Comment (RemieRichards) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: It does make sense (to me atleast), ```current.X``` is only read for that one if statement, and prior to the fix it used to be the cumulative amount of Spacing/Padding, where as now it's one element's worth of Padding, and then reset to 0, I can remove the comment but I felt just setting it to 0 *looked* dodgy enough to need explaining.

File: osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs (hunk 1 above)
Comment (RemieRichards) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: Doesn't appear that way in the test.

Comment (peppy) [inline osu.Framework.VisualTests/Tests/TestCaseScrollableFlow.cs]: I haven't looked at the test. I'm saying you changed the display logic. You need to understand that and not just be "oh it looks roughly the same"

File: osu.Framework/Graphics/Containers/FlowContainer.cs (hunk 4 above)
Comment (peppy) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: read the logic again. current is used in a cumulative fashion to position *each element* in the flow. it is not only used in the if statement.

Comment (RemieRichards) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: Yes, it used to be cumulative, however now when it enters this if, it no longer is, as it gets reset to 0 (then ```Size.X``` is added on further down to make it 1 count of ```Size.X```)

File: osu.Framework/Graphics/Containers/FlowContainer.cs [hunk 6]
```diff
@@ -94,7 +97,7 @@ protected override void UpdateLayout()
 
                     if (Children.FirstOrDefault() == null) return Vector2.Zero;
 
-                    Vector2 current = new Vector2(Math.Max(0, ElementPadding.X), Math.Max(0, ElementPadding.Y));
+                    Vector2 current = new Vector2(0, 0);
```
Comment (peppy) [inline osu.Framework/Graphics/Containers/FlowContainer.cs]: Vector2.Zero

**Pull request merged**